src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -422,473 +388,307 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
-            this.currentOperationContentId = null;
-        }
+        /// <summary>
+        /// Ends an active changeset.
+        /// </summary>
+        protected abstract void WriteEndChangesetImplementation();
 
         /// <summary>
-        /// Verifies that calling CreateOperationRequestMessage if valid.
+        /// Creates an <see cref="ODataBatchOperationResponseMessage"/> for writing an operation of a batch response.
         /// </summary>
-        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
-        /// <param name="method">The Http method to be used for this request operation.</param>
-        /// <param name="uri">The Uri to be used for this request operation.</param>
         /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
-        private void VerifyCanCreateOperationRequestMessage(bool synchronousCall, string method, Uri uri, string contentId)
-        {
-            this.ValidateWriterReady();
-            this.VerifyCallAllowed(synchronousCall);
-
-            if (this.rawOutputContext.WritingResponse)
-            {
-                this.ThrowODataException(Strings.ODataBatchWriter_CannotCreateRequestOperationWhenWritingResponse);
-            }
+        /// <returns>The message that can be used to write the response operation.</returns>
+        protected abstract ODataBatchOperationResponseMessage CreateOperationResponseMessageImplementation(string contentId);
 
-            ExceptionUtils.CheckArgumentStringNotNullOrEmpty(method, "method");
-            if (this.changeSetBoundary != null)
-            {
-                if (HttpUtils.IsQueryMethod(method))
-                {
-                    this.ThrowODataException(Strings.ODataBatch_InvalidHttpMethodForChangeSetRequest(method));
-                }
+        /// <summary>
+        /// Writes all the pending headers and prepares the writer to write a content of the operation.
+        /// </summary>
+        protected abstract void StartBatchOperationContent();
 
-                if (string.IsNullOrEmpty(contentId))
-                {
-                    this.ThrowODataException(Strings.ODataBatchOperationHeaderDictionary_KeyNotFound(ODataConstants.ContentIdHeader));
-                }
-            }
+        /// <summary>
+        /// Disposes the batch writer and set the 'OperationStreamRequested' batch writer state;
+        /// called after the flush operation(s) have completed.
+        /// </summary>
+        protected abstract void DisposeBatchWriterAndSetContentStreamRequestedState();
 
-            ExceptionUtils.CheckArgumentNotNull(uri, "uri");
-        }
+        /// <summary>
+        /// Verifies that the writer is in correct state for the Flush operation.
+        /// </summary>
+        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
+        protected abstract void VerifyCanFlush(bool synchronousCall);

```

Comment:
>VerifyCanFlush [](start = 32, length = 14)

See ODataWriterCore for a slightly different factoring.  For consistency, we should make VerifyCanFlush private (as it was before), and have the implementation in the base class call VerifyNotDisposed (which should be protected) and then call VerifyCallAllowed.

Reply:
Original VerifyCanFlush() was private because its was called from Flush() which use only one context type. Now with the additional JsonLight context type, Flush() need to be abstract in base class so that derived class can invoke, e.g. this.RawOutputContext.Flush() (note that Flush() is not available in common base ODataOutputContext). Therefore making VerifyCanFlush() private won't work for Flush().

---
In reply to: [140934183](https://github.com/OData/odata.net/pull/960#discussion_r140934183) [](ancestors = 140934183)

Reply:
Yes, you're right. If we implement Flush in the base class, we can call VerifyCanFlush in the base class, then call protected FlushSynchrously (or equivalent) in the derived class. VerifyCanFlush should be implemented in the base class and call VerifyNotDisposed, which should be protected and overridden in the derived class.  The rest of the logic in VerifyCanFlush can be in the base class.  VerifyCallAllowed should then be made private. That would mean that 1) if we needed to add logic to Flush in the base class later we could, 2) we would ensure that VerifyCanFlush is always called before flush, and 3) it would be more consistent with other writer implementations.

---
In reply to: [142566060](https://github.com/OData/odata.net/pull/960#discussion_r142566060) [](ancestors = 142566060,140934183)